ambry-api/src/main/java/com.github.ambry/router/AsyncWritableChannel.java

```diff
@@ -44,5 +48,60 @@
    * @return a {@link Future} that will eventually contain the result of the write operation (the number of bytes
    *         written).
    */
-  public Future<Long> write(ByteBuffer src, Callback<Long> callback);
+  Future<Long> write(ByteBuffer src, Callback<Long> callback);
+
+  /**
+   * Write data in {@code src} to the channel and the {@code callback} will be invoked once the write succeeds or fails.
+   * This method is the counterpart for {@link ByteBuf}. It shares the same guarantee as the {@link #write(ByteBuffer, Callback)}.
+   * Whoever implements this interface, shouldn't release the {@code src}. If releasing is expected after finishing writing
+   * to channel, please releas this {@link ByteBuf} in the callback method.
+   * @param src The data taht needs to be written to the channel.
+   * @param callback The {@link Callback} that will be invoked once the write succeeds/fails. This can be null.
+   * @return a {@link Future} that will eventually contain the result of the write operation (the number of bytes
+   *         written).
+   */
+  default Future<Long> write(ByteBuf src, Callback<Long> callback) {
```

Comment:
Could you write a unit test for this default method to check that it works as expected in all of the 3 cases nioBufferCount cases?

Reply:
the case where nioBufferCount is -1 is pretty hard to construct so I will not do that. And I have unit tests for CopyingAsyncWritableChannel, which didn't have it's owner implementation of write(ByteBuf), that would serve to test the write(ByteBuf) as default method.